Make Msaa a component not a resource.

A-Walrus opened the pull request:
This way it can be configured per camera.

# Objective
Make `Msaa` a component so that it can be configured per camera
Fixes #7194 
## Solution

- Change `Msaa` to a component, which is included in the camera bundles, and gets extracted in the `extract_cameras` system.
- Add `Msaa` component to camera queries, instead of getting it as a resource.
- Migrate the `Msaa` related examples to new system.

---

## Changelog

### Changed
Msaa is now a component and not a resource, to allow per-camera configuration.

## Migration Guide

In order to access `Msaa` options, use a query for the `Msaa` component which is now part of the camera bundles, instead of using a `Res(Mut)<Msaa>`.
If using a camera without a bundle, make sure to add the `Msaa` component yourself!

DasLixou reviewed (changes requested):
When the `Camera3dBundle` already has msaa field, why do we need to insert it again? Also: is there a way to make the msaa field in the bundle an option? I don’t think everyone want it. A way to disable it by just saying:
```rust
        msaa: None,
        ..default()
    }
```
would be quite handy

alice-i-cecile commented:
> Also: is there a way to make the msaa field in the bundle an option

Unfortunately no, see #2157.

DasLixou commented:
> > Also: is there a way to make the msaa field in the bundle an option
> 
> Unfortunately no, see #2157.

That’s sad. So I see two options:
* Let the msaa be in the bundle and cleanup the examples
* Remove the msaa field and let the example as it is

A-Walrus commented:
> When the `Camera3dBundle` already has msaa field, why do we need to insert it again?

My bad, it was unnecessary, cleaned up now.

MinerSebas commented:
You disable MSAA, by setting samples to 1.
The best course of action would probably be to merge #6954, which makes it more clear how to disable it.

DasLixou commented:
> You disable MSAA, by setting samples to 1.
> The best course of action would probably be to merge #6954, which makes it more clear how to disable it.

Like it 👍 . Yep. Then my request can be marked as resolved

IceSentry commented on crates/bevy_core_pipeline/src/core_2d/camera_2d.rs:
```diff
@@ -77,6 +78,7 @@ impl Camera2dBundle {
             camera: Camera::default(),
             camera_2d: Camera2d::default(),
             tonemapping: Tonemapping::Disabled,
+            msaa: Msaa { samples: 1 },
```
This should use the default value

IceSentry replied:
```suggestion
            msaa: Msaa::default(),
```

IceSentry commented on crates/bevy_core_pipeline/src/core_3d/camera_3d.rs:
```diff
@@ -85,6 +86,7 @@ impl Default for Camera3dBundle {
             transform: Default::default(),
             global_transform: Default::default(),
             camera_3d: Default::default(),
+            msaa: Msaa { samples: 1 },
```
```suggestion
            msaa: Msaa::default(),
```

IceSentry commented on examples/3d/msaa.rs:
```diff
@@ -40,12 +39,14 @@ fn setup(
     // camera
     commands.spawn(Camera3dBundle {
         transform: Transform::from_xyz(-3.0, 3.0, 5.0).looking_at(Vec3::ZERO, Vec3::Y),
+        msaa: Msaa::default(),
```
This is the default value, but I think this should show the full struct like the original example

```suggestion
        msaa: Msaa { sample: 4 },
```

IceSentry reviewed (changes requested):
Couple of nits.

The implementation looks good, I'm not entirely sure this is necessary, but I'll approve it once my comments are resolved.

nicopap commented:
Note that this change opens up a new footgun, if the user has two different sampling rates for MSAA on the same target, it will panic, here is a minimal reproducible example:

```rust
use bevy::{core_pipeline::clear_color::ClearColorConfig, prelude::*, render::view::RenderLayers};

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_startup_system(startup)
        .run();
}

fn startup(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.spawn(Camera2dBundle {
        msaa: Msaa { samples: 4 },
        camera_2d: Camera2d {
            clear_color: ClearColorConfig::Custom(Color::WHITE),
        },
        ..default()
    });
    commands.spawn((
        Camera2dBundle {
            msaa: Msaa { samples: 1 },
            camera_2d: Camera2d {
                clear_color: ClearColorConfig::None,
            },
            camera: Camera {
                order: 1,
                ..default()
            },
            ..default()
        },
        RenderLayers::layer(1),
    ));

    commands.spawn((SpriteBundle {
        texture: asset_server.load("branding/icon.png"),
        transform: Transform::from_scale(Vec3::splat(2.0)),
        ..default()
    },));
    commands.spawn((
        SpriteBundle {
            texture: asset_server.load("branding/icon.png"),
            ..default()
        },
        RenderLayers::layer(1),
    ));
}
```

A-Walrus commented:
@nicopap I added a check that detects and warns when multiple cameras have different `Msaa`s on the same target. Also added a note in the doc comment about the footgun.

robtfm commented:
> if the user has two different sampling rates for MSAA on the same target, it will panic

this is a good point, and makes me think the `Camera` entity is the wrong place for the MSAA component. 

I was about to suggest it should be on the `RenderTarget` instead, but actually there's nothing to stop people making multiple `RenderTarget`s with the same underlying target image or window so it's the same problem. so actually it should be on the `Window` and the `Image` themselves.

`Image` already has this info, (`image.texture_descriptor.sample_count`), so i guess we should use that instead of the resource when looking up msaa for a camera with a `RenderTarget::Image` target. For windows, perhaps we should wait until we have #5589 before making any changes.
